Mathlib/Algebra/Category/ModuleCat/Basic.lean

```diff
@@ -88,27 +71,121 @@ instance : CoeSort (ModuleCat.{v} R) (Type v) :=
+
+/- Provided for rewriting. -/
+lemma id_apply (M : ModuleCat.{v} R) (x : M) :
+    (𝟙 M : M ⟶ M) x = x := by simp
+
+@[simp]
+lemma hom_comp {M N O : ModuleCat.{v} R} (f : M ⟶ N) (g : N ⟶ O) :
+    (f ≫ g).hom = g.hom.comp f.hom := rfl
+
+/- Provided for rewriting. -/
+lemma comp_apply {M N O : ModuleCat.{v} R} (f : M ⟶ N) (g : N ⟶ O) (x : M) :
+    (f ≫ g) x = g (f x) := by simp
+
+@[ext]
+lemma hom_ext {M N : ModuleCat.{v} R} {f g : M ⟶ N} (hf : f.hom = g.hom) : f = g :=
+  Hom.ext hf
+
+lemma hom_bijective {M N : ModuleCat.{v} R} :
+    Function.Bijective (Hom.hom : (M ⟶ N) → (M →ₗ[R] N)) where
+  left f g h := by cases f; cases g; simpa using h
+  right f := ⟨⟨f⟩, rfl⟩
+
+/-- Typecheck a `LinearMap` as a morphism in `ModuleCat R`. -/
+abbrev asHom {X Y : Type v} [AddCommGroup X] [Module R X] [AddCommGroup Y] [Module R Y]
+    (f : X →ₗ[R] Y) : of R X ⟶ of R Y :=
+  ⟨f⟩
+
+@[deprecated (since := "2024-10-06")] alias ofHom := ModuleCat.asHom
+
+lemma hom_asHom {X Y : Type v} [AddCommGroup X] [Module R X] [AddCommGroup Y]
+    [Module R Y] (f : X →ₗ[R] Y) : (asHom f).hom = f := rfl
 
-instance {M N : ModuleCat.{v} R} : FunLike (M ⟶ N) M N :=
-  LinearMap.instFunLike
+@[simp]
+lemma asHom_hom {M N : ModuleCat.{v} R} (f : M ⟶ N) :
+    asHom (Hom.hom f) = f := rfl
+
+@[simp]
+lemma asHom_id {M : Type v} [AddCommGroup M] [Module R M] : asHom LinearMap.id = 𝟙 (of R M) := rfl
+
+@[simp]
+lemma asHom_comp {M N O : Type v} [AddCommGroup M] [AddCommGroup N] [AddCommGroup O] [Module R M]
+    [Module R N] [Module R O] (f : M →ₗ[R] N) (g : N →ₗ[R] O) :
+    asHom (g.comp f) = asHom f ≫ asHom g :=
+  rfl
 
-instance {M N : ModuleCat.{v} R} : LinearMapClass (M ⟶ N) R M N :=
-  LinearMap.semilinearMapClass
+lemma asHom_apply {M N : Type v} [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+    (f : M →ₗ[R] N) (x : M) : asHom f x = f x := rfl
+
+@[simp]
+lemma inv_hom_apply {M N : ModuleCat.{v} R} (e : M ≅ N) (x : M) : e.inv (e.hom x) = x := by
+  rw [← comp_apply]
+  simp
+
+@[simp]
+lemma hom_inv_apply {M N : ModuleCat.{v} R} (e : M ≅ N) (x : N) : e.hom (e.inv x) = x := by
+  rw [← comp_apply]
+  simp
+
+/-- `ModuleCat.Hom.hom` bundled as an `Equiv`. -/
+def homEquiv {M N : ModuleCat.{v} R} : (M ⟶ N) ≃ (M →ₗ[R] N) where
+  toFun := Hom.hom
+  invFun := asHom
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+end
+
+instance : Inhabited (ModuleCat R) :=
+  ⟨of R R⟩
 
 instance moduleConcreteCategory : ConcreteCategory.{v} (ModuleCat.{v} R) where
   forget :=
     { obj := fun R => R
-      map := fun f => f.toFun }
-  forget_faithful := ⟨fun h => LinearMap.ext (fun x => by
-    dsimp at h
-    rw [h])⟩
+      map := fun f => f.hom }
+  forget_faithful := ⟨fun h => by ext x; simpa using congrFun h x⟩
+
+lemma forget_obj {M : ModuleCat.{v} R} : (forget (ModuleCat.{v} R)).obj M = M := rfl
+
+lemma forget_map {M N : ModuleCat.{v} R} (f : M ⟶ N) :
+    (forget (ModuleCat.{v} R)).map f = f :=
+  rfl
```

Comment:
I think indeed these shouldn't be `simp`, but let's say so.

Reply:
@chrisflav, am I correct that you removed these from the `simp` set in 6c554e33a35498ea9ef0fc288523c26d980e4fcc because:

* `forget_obj` will apply as a `dsimp` lemma to the (co)domain of morphisms, causing definitional equality issues?
* `forget_map` has a morphism on the LHS but a plain function on the RHS?

Reply:
But I also agree with your two points.